File: android/app/src/main/java/ai/willkim/wkwhisperkey/FineTuneWorker.java
package ai.willkim.wkwhisperkey;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;
import androidx.work.WorkManager;
import androidx.work.OneTimeWorkRequest;

import java.util.List;

public class FineTuneWorker extends Worker {

    public FineTuneWorker(@NonNull Context context, @NonNull WorkerParameters params) {
        super(context, params);
    }

    @NonNull
    @Override
    public Result doWork() {
        List<WhisperData> samples = WhisperStore.sampleBatch(8);
        for (WhisperData data : samples) {
            // pass data.wave, data.text
        }
        new WhisperEngine().finetune(samples);
        return Result.success();
    }

    public static void enqueue(Context ctx) {
        WorkManager.getInstance(ctx)
                .enqueue(new OneTimeWorkRequest.Builder(FineTuneWorker.class).build());
    }
}

File: android/app/src/main/java/ai/willkim/wkwhisperkey/FineTuneWorker.kt
class FineTuneWorker(ctx: Context, params: WorkerParameters)
    : CoroutineWorker(ctx, params) {

    override suspend fun doWork(): Result {
        val samples = WhisperStore.sampleBatch(8)
        val pairs = samples.map { it.wave to it.text }
        WhisperEngine().finetune(pairs)
        return Result.success()
    }

    companion object {
        fun enqueue(ctx: Context) {
            WorkManager.getInstance(ctx).enqueue(
                OneTimeWorkRequestBuilder<FineTuneWorker>().build()
            )
        }
    }
}
